### ISSUE TITLE
Replace for loop in _build() with tf.while_loop for EncodeProcessDecode class

### ISSUE THREAD

**Issue by ferreirafabio** [deepmind/graph_nets#57]
I'm trying to transfer the `_build()` function of `EncodeProcessDecode` into a `tf.while_loop` variant for variable-sized `num_processing_steps` that I now feed as a placeholder value at every build() call.
So far, I'm struggling to concatenate the `decoded_op` in a list because I can only come up with a version using `shape_invariants` which in turn requires to know shapes of `GraphTuple`. However, calls like `tf.shape(output_op)` won't work, nor does `tf.concat` accept values other than Tensors. To my knowledge, a variant not requiring `shape_invariants` must rely on `tf.TensorShape()` which is only defined for TensorFlow primitive data types.

Can someone give me a hint how to use `tf.while_loop` for replacing the _build() Python for loop? Can one use `utils_tf.concat` for the while_loop iterations instead of `tf.concat`? How should its output be further processed or how can the concatenated GraphTuple elements be accessed for constructing the loss function again? An example would very handy! Here's my approach so far:

```
def _build(self, input_op, input_ctrl_ph, num_processing_steps, is_training):
        latent = self._encoder(input_op, is_training)
    
        latent_global = self._encoder_globals(input_op, is_training)
        latent_init = latent.replace(globals=latent_global.globals)
    
        global_T = self._encoder_globals(input_ctrl_ph, is_training).globals

        i = tf.constant(0)
        output_ops = tf.Variable([])
    
        def iter(i, *args):
            return i < num_processing_steps

        def body(i, output_ops, latent):
            global_t = tf.gather(global_T, i)
            global_t = tf.expand_dims(global_t, 0)
            latent = latent.replace(globals=global_t)
        
            latent = self._core(latent)
        
            decoded_op = self._decoder(latent, is_training)
            return i+1, tf.concat([output_ops, decoded_op], axis=0), latent

        i, outputs, latent = tf.while_loop(iter, body, [i, output_ops, latent_init], 
                             shape_invariants=[i.get_shape(), 
                                               tf.TensorShape([None]), 
                                               tf.TensorShape([None])])
```

**Comment by alvarosg** [comment]
It is usually easier to use `tf.TensorArray`s rather than concatenation. Also, for now you will need to use separate tensor arrays for node, edges and globals explicitly.

**Comment by ferreirafabio** [comment]
Thanks!
